### PR TITLE
Update `qml.math.allclose` to work for trainable Torch tensors (motivated by `qml.equal`)

### PR DESCRIPTION
**Pull request by antalszava**
**Context:**
```py
pl_tensor = qml.numpy.array(0.3, requires_grad=True)
torch_tensor = torch.tensor(0.3, requires_grad=True)
qml.math.allclose(pl_tensor, torch_tensor)
```
Raises
```py
RuntimeError: Can't call numpy() on Tensor that requires grad. Use tensor.detach().numpy() instead.
```
**Description of the Change:**
Updates `qml.math.allclose` to catch `RuntimeError`. This way trainable Torch tensors are correctly converted to NumPy.

**Benefits:**
`qml.math.allclose` works with Torch Tensors that have `requires_grad=True` set.

**Possible Drawbacks:**
N/A

**Related GitHub Issues:**
N/A

### CODE REVIEW

File: tests/ops/functions/test_equal.py [hunk 1]
```diff
@@ -834,9 +834,80 @@ def test_equal_simple_op_remaining(self):
             check_interface=False,
         )
 
-    def test_equal_with_different_arithmetic_depth(self):
-        """Test equal method with two operators with different arithmetic depth."""
-        assert not qml.equal(qml.adjoint(qml.PauliX(0)), qml.adjoint(qml.adjoint(qml.PauliX(0))))
```
Comment (AlbertMitjans) [inline tests/ops/functions/test_equal.py:839]: Why was this test deleted?

File: tests/ops/functions/test_equal.py [hunk 2]
```diff
@@ -834,9 +834,80 @@ def test_equal_simple_op_remaining(self):
...
+        # array into a Tracer
+        def jax_assertion_func(x, y):
+            op1 = qml.RY(jax.numpy.array(x), wires=1)
+            op2 = qml.RY(torch.tensor(y, requires_grad=True), wires=1)
+            assert qml.equal(op1, op2, check_interface=False, check_trainability=True)
+            return x
+
+        par = 0.3
+        for tensor in non_jax_tensors:
+            jax.grad(jax_assertion_func, argnums=0)(par, par)
```
Comment (AlbertMitjans) [inline tests/ops/functions/test_equal.py:880]: `tensor` is not used

Comment (AlbertMitjans) [inline tests/ops/functions/test_equal.py:880]: `tensor` is not used. Maybe you wanted to do this?

```suggestion
            jax.grad(jax_assertion_func, argnums=0)(tensor, par)
```

**Review by AlbertMitjans — COMMENTED** [review]
Nice! 💯 Left some comments.

File: tests/ops/functions/test_equal.py [hunk 3]
```diff
@@ -834,9 +834,80 @@ def test_equal_simple_op_remaining(self):
...
+        -Autograd and Torch
+        """
+        import jax
+        import tensorflow as tf
+        import torch
+
+        wire = 0
+
+        pl_tensor = qml.numpy.array(0.3, requires_grad=True)
+        jax_array = jax.core.Tracer(0.3)
```
Comment (AlbertMitjans) [inline tests/ops/functions/test_equal.py:861]: Unused variable.

File: tests/ops/functions/test_equal.py (hunk 1 above)
Comment (rmoyard) [reply tests/ops/functions/test_equal.py:839]: Same as Albert.

File: tests/ops/functions/test_equal.py (hunk 2 above)
Comment (rmoyard) [reply tests/ops/functions/test_equal.py:880]: Yes I guess it should be passed in the `jax_assertion_func` because currently it only test torch.

**Review by rmoyard — COMMENTED** [review]
Great catch @antalszava 👍  think one of the test should be slightly changed.

File: tests/ops/functions/test_equal.py (hunk 1 above)
Comment (antalszava) [reply tests/ops/functions/test_equal.py]: Oh by mistake, thanks for catching this!

File: tests/ops/functions/test_equal.py (hunk 2 above)
Comment (antalszava) [reply tests/ops/functions/test_equal.py]: Yep, thanks for catching this!

**Comment by antalszava** [comment]
Thank you @AlbertMitjans @rmoyard for the comments and catches! Updated as per your feedback.

**Review by rmoyard — APPROVED** [review]
Thanks @antalszava for the changes 💯 approved!

**Review by AlbertMitjans — APPROVED** [review]
🎉

**Pull request merged**